[InstCombine] optimize powi(X,Y) * X with Ofast

vfdff opened the pull request:
Try to transform the powi(X, Y) * X into powi(X, Y+1) with Ofast

For this case, when the Y is 3, then powi(X, 4) is replaced by X2 = X * X; X2 * X2 in the further step.
Similar to D109954, who requires reassoc.

Fixes https://github.com/llvm/llvm-project/issues/69862.

vfdff committed:
[InstCombine] optimize powi(X,Y) * X with Ofast

Try to transform the powi(X, Y) * X into powi(X, Y+1) with Ofast

For this case, when the Y is 3, then powi(X, 4) is replaced by
X2 = X * X; X2 * X2 in the further step.
Similar to D109954, who requires reassoc.

Fixes https://github.com/llvm/llvm-project/issues/69862.

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -716,6 +716,18 @@ Instruction *InstCombinerImpl::visitFMul(BinaryOperator &I) {
 
+    // powi(X, Y) * X --> powi(X, Y+1)
+    // X * powi(X, Y) --> powi(X, Y+1)
+    if (match(&I, m_c_FMul(m_OneUse(m_Intrinsic<Intrinsic::powi>(m_Value(X),
```
The reassoc check is really hard to find, this whole if (reassoc) should probably be split to a helper function.

Also, should check the reason on the inner operation too?

vfdff replied:
hi @arsenm , do you mean both the operation `fmul `and `powi `should have the `reassoc ` ?
  I find all the past combines only check the instruction `I` itself only.

arsenm replied:
I consider that to be a bug. I think we should have strict flag requirements for both participants

vfdff replied:
Fixed with new commit 4423e80cb, thanks

arsenm replied:
This commit isn't upstream, or in a PR? Did you mean to link a PR with it?

vfdff replied:
oh, sorry, please ignore that. I have rebase it on this PR (commit 86b9575) because C/C++ code formatter, and forget to update the reply.

nikic reviewed (commented):
We have three variants of this optimization now: `powi(x,a)*powi(x,b) => powi(x,a+b)`, `powi(x,a)/x => powi(x,a-1)` and `powi(x,a)*x => powi(x,a+1)`.

I'd suggest to move most of the transform into a helper function, that will accept `x, a, b`, `x, a, -1` and `x, a, 1` in the above cases and then do the common checks and transform.

This will also fix the bug in the current `powi * powi` transform, which fails to perform the necessary overflow check.

arsenm commented:
The ARM case already does end up with this optimization? How did that happen?

vfdff commented:
> We have three variants of this optimization now: `powi(x,a)*powi(x,b) => powi(x,a+b)`, `powi(x,a)/x => powi(x,a-1)` and `powi(x,a)*x => powi(x,a+1)`.
> 
> I'd suggest to move most of the transform into a helper function, that will accept `x, a, b`, `x, a, -1` and `x, a, 1` in the above cases and then do the common checks and transform.
> 
> This will also fix the bug in the current `powi * powi` transform, which fails to perform the necessary overflow check.

Three variants of this optimization have many differences except the pattern match form,  and it seems the only common check is `hasAllowReassoc()` for instruction `I`?
1、`powi(X, Y) / X --> powi(X, Y-1)` need  `hasAllowReassoc()` and `hasNoNaNs()` for both` I and powi(X, Y)` , called in visitFDiv, check 
2、`powi(X, Y) * X --> powi(X, Y+1)` need `hasAllowReassoc()`  for both `I and powi(X, Y)` , called in visitFMul
3、`powi(x, y) * powi(x, z) -> powi(x, y + z)` need `hasAllowReassoc()` for all `I、powi(X, Y) and powi(X, Z)` , called in visitFMul

vfdff commented:
hi @arsenm, @nikic : I have another idea, not sure if it's appropriate?
-  According to above discussion, many other optimizations have similar problems due to the need to recursively determine the operand `fmt attributes` involved. But in fact, the problem has not been exposed for so long. Is it because these `fmt attributes` are set in units of at least one function, that is, IR operands in the same function should be consistent? That's why this problem has not been exposed for so long?
- If so, then we can just keep the same as we are now.

jcranmer-intel commented:
>     * According to above discussion, many other optimizations have similar problems due to the need to recursively determine the operand `fmt attributes` involved. But in fact, the problem has not been exposed for so long. Is it because these `fmt attributes` are set in units of at least one function, that is, IR operands in the same function should be consistent? That's why this problem has not been exposed for so long?

Fast-math flags tend to be applied on a per-project basis via command-line (i.e., per-IR module) flags, so it tends to be very rare for the flags to differ within a function, unless you're doing LTO, which is why it's unlikely to crop up very frequently.

vfdff committed:
[InstCombine] create a helper function foldPowiReassoc, NFC

vfdff committed:
[InstCombine] Add restrict reassoc for the operands of fmul

According the discussion, except the fmul itself, all its operands
should also have reassoc flag.
Add new API m_AllowReassoc to check reassoc flag

vfdff commented:
> > ```
> > * According to above discussion, many other optimizations have similar problems due to the need to recursively determine the operand `fmt attributes` involved. But in fact, the problem has not been exposed for so long. Is it because these `fmt attributes` are set in units of at least one function, that is, IR operands in the same function should be consistent? That's why this problem has not been exposed for so long?
> > ```
> 
> Fast-math flags tend to be applied on a per-project basis via command-line (i.e., per-IR module) flags, so it tends to be very rare for the flags to differ within a function, unless you're doing LTO, which is why it's unlikely to crop up very frequently.

Thanks, I add **the restrict for the operands of fmul**.

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -571,6 +571,55 @@ Instruction *InstCombinerImpl::foldFPSignBitOps(BinaryOperator &I) {
+  Value *X, *Y, *Z;
+
+  // Make sure all operands have reassoc flag if they are powi.
+  if (!all_of(I.operands(), [](Value *V) {
+        if (match(V, m_Intrinsic<Intrinsic::powi>(m_Value(), m_Value()))) {
+          Instruction *Powi = cast<Instruction>(V);
+          return Powi->hasAllowReassoc();
+        }
+        return true;
+      }))
+    return nullptr;
```
I think it would be easier to follow if you checked the reassoc flag below where you're already matching the actual powi call, instead of this pre-check of the operands

vfdff replied:
Apply your comment, thanks

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -571,6 +571,55 @@ Instruction *InstCombinerImpl::foldFPSignBitOps(BinaryOperator &I) {
+    Value *YZ;
+    InstCombiner::BuilderTy &Builder = IC.Builder;
+
+    if (auto *C = dyn_cast<ConstantInt>(Z)) {
+      if (C->isOne())
+        YZ = Builder.CreateAdd(Y, ConstantInt::get(Y->getType(), 1));
+    } else
+      YZ = Builder.CreateAdd(Y, Z);
```
I don't understand this YZ logic; if you can just do CreateAdd for an arbitrary add you can do it for the constant case. As-is this will by use of undefined value for constants that aren't 1

vfdff replied:
Yes, It need be fix. I check the const one before because I only have a call site with that value, but indeed it doesn't need restrict for the const value.

vfdff commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -571,6 +571,55 @@ Instruction *InstCombinerImpl::foldFPSignBitOps(BinaryOperator &I) {
+                                                               m_Value(Y))),
+                         m_Deferred(X))) &&
+      willNotOverflowSignedAdd(Y, ConstantInt::get(Y->getType(), 1), I))
+    return createPowiExpr(I, *this, X, Y, ConstantInt::get(Y->getType(), 1));
```
fixed,thanks

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -571,6 +571,50 @@ Instruction *InstCombinerImpl::foldFPSignBitOps(BinaryOperator &I) {
+                         m_Deferred(X)))) {
+    Constant *One = ConstantInt::get(Y->getType(), 1);
+    if (willNotOverflowSignedAdd(Y, One, I)) {
+      auto *Powi = dyn_cast<IntrinsicInst>(I.getOperand(0));
+      if (!Powi)
+        Powi = cast<IntrinsicInst>(I.getOperand(1));
+      if (Powi->hasAllowReassoc())
```
We should really have a way to put this flag check inside the matcher

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -571,6 +571,50 @@ Instruction *InstCombinerImpl::foldFPSignBitOps(BinaryOperator &I) {
 }
 
+Instruction *InstCombinerImpl::foldPowiReassoc(BinaryOperator &I) {
+  Value *X, *Y, *Z;
```
Sink these down where they are used, looks like shadowing now

arsenm commented on llvm/test/Transforms/InstCombine/powi.ll:
```diff
@@ -192,42 +225,42 @@ define double @powi_fmul_powi_same_power(double %x, i32 %y, i32 %z) {
+  %p1 = tail call reassoc double @llvm.powi.f64.i32(double %x, i32 %z)
   tail call void @use(double %p1)
-  %p2 = tail call double @llvm.powi.f64.i32(double %x, i32 %y)
+  %p2 = tail call reassoc double @llvm.powi.f64.i32(double %x, i32 %y)
```
If we don't have a test where the 2 powi declarations have different integer types, should add one

arsenm commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -572,37 +572,35 @@ Instruction *InstCombinerImpl::foldFPSignBitOps(BinaryOperator &I) {
-                         m_Deferred(X))) &&
-      willNotOverflowSignedAdd(Y, ConstantInt::get(Y->getType(), 1), I))
-    return createPowiExpr(I, *this, X, Y, ConstantInt::get(Y->getType(), 1));
+  if (match(&I, m_c_FMul(m_AllowReassoc(m_OneUse(m_Intrinsic<Intrinsic::powi>(
```
the m_AllowReassoc should be paired with the m_Intrinsic part, not after the oneUse

arsenm commented on llvm/include/llvm/IR/PatternMatch.h:
```diff
@@ -68,6 +68,22 @@ template <typename T> inline OneUse_match<T> m_OneUse(const T &SubPattern) {
+};
+
+template <typename T>
+inline AllowReassoc_match<T> m_AllowReassoc(const T &SubPattern) {
```
This isn't quite the API I had in my mind. I envisioned the required flags as a template parameter to the existing matchers, so you could have something like: 
`m_FMul<Reassoc>(), m_Intrinsic<powi, Reassoc>`

However I wasn't expecting you to do anything for this in this patch. This is fine for now
